### PR TITLE
This commit includes the following features: Updated CRT-PMT matching…

### PR DESCRIPTION
**Pull request by francescopoppi**
This PR is in parallel with: https://github.com/SBNSoftware/icaruscode/pull/539 (approved by the reviewers)

This PR contains 3 major elements:

1. An updated and improved CRT-PMT matching analysis tool (docDB 30190).
2. A bug fix that prevents stage0 failing with Runs after run number 9580. Starting from these runs, two additional spare Top CRT modules were included in the DAQ. The bug was due to the CRT decoder stage failing to find the cable delay of the 2 new modules since they were not present in the delay definitions. This patch includes the corresponding delays and prevents the failure of the Stage0 processing **_but_** the CRT hits coming from these two external modules will not be reconstructed since they cannot be matched with a module in the gdml geometry of ICARUS. I believe we do not want to include them in the simulation/gdml since they potentially will move around to sample different region of the detector roof (above concrete). There is a branch forked from develop where I am (succesfully) testing the reconstruction of these two modules to have the correspondent CRT hits available in the CRTHits data products, but more testing is foreseen. I expect at least one month of testing before releasing this feature, but nonetheless it is definitely not critical.
3. Previously the light propagation delay inside the WLS fibre was determined by using the light group velocity in the fibre, now we are using a data driven model that depends only by the triggering sector. See DodcDB 29504 slide 9. 

NOTE: The Top CRT modules can be represented as 64 triggering sector, depending only on the quadruplets of triggering channel.

### CODE REVIEW

**Comment by mmrosenberg** [comment]
trigger build LArSoft/lar*@LARSOFT_SUITE_v09_63_00 SBNSoftware/sbncode@v09_63_00_02 SBNSoftware/sbnanaobj@v09_20_00_01 SBNSoftware/sbndaq_artdaq_core@v1_04_02of0 SBNSoftware/sbnobj@v09_15_00_01 SBNSoftware/icarus_signal_processing@v09_63_00_02 SBNSoftware/icarusalg@v09_63_00_02p03 SBNSoftware/icarusutil@v09_62_00

**Review by PetrilloAtWork — CHANGES_REQUESTED** [review]
I have left a lot of coding comments that may be considered by the author (actually I suspect some of these pertain parts of code inherited from the older version...).
But I request two changes:
* the threshold value `400` be turned into a configurable parameter
* fix the bug highlighted in the comment

--- 

More in general, I feel like here this is material for algorithm code, not just an analyzer module.
This matching is useful enough that we'll want it encapsulated probably into a producer, and its result serialised into a data product. I am not sure how far we are from there.
Thus I have been pickier than my usual for analyzer modules.

**Comment by francescopoppi** [comment]
Right... as @PetrilloAtWork mentioned I forgot to merge comments addressed in previous (couple of months ago) PR which are included here: https://github.com/SBNSoftware/icaruscode/blob/develop/icaruscode/Filters/FilterCRTPMTMatching_module.cc
Today I will merge them all into this PR (and in the one for develop)

File: icaruscode/CRT/CRTPMTMatchingAna_module.cc [hunk 1]
```diff
@@ -174,319 +322,391 @@ icarus::crt::CRTPMTMatchingAna::CRTPMTMatchingAna(fhicl::ParameterSet const& p)
...
+      vector<art::Ptr<recob::OpHit>> hits = findManyHits.at(iflash);
+      int nPMTsTriggering = 0;
+      double firstTime=999999;
+      double flash_pos[3]={0,0,0};
+      double ampsum=0, t_m=0;   
+      for(auto const& hit : hits) {
+        if (hit->Amplitude() > 400) nPMTsTriggering++;
+        if(firstTime > hit->PeakTime()) firstTime=hit->PeakTime();
+        double pos[3];
+        fGeometryService->OpDetGeoFromOpChannel(hit->OpChannel()).GetCenter();
```
[the remark this answers is not in the document]
Comment (francescopoppi) [reply icaruscode/CRT/CRTPMTMatchingAna_module.cc:482]: well... This is awkward. This is probably why I was getting weird values on the OpFlash barycenter.

File: icaruscode/CRT/CRTPMTMatchingAna_module.cc [hunk 2]
```diff
@@ -174,319 +322,391 @@ icarus::crt::CRTPMTMatchingAna::CRTPMTMatchingAna(fhicl::ParameterSet const& p)
...
+        double amp=hit->Amplitude();
+        ampsum=ampsum+amp;
+        fOpHitX.push_back(pos[0]);
+        fOpHitY.push_back(pos[1]);
+        fOpHitZ.push_back(pos[2]);
+        fOpHitT.push_back(hit->PeakTime());
+        fOpHitA.push_back(amp);
+        flash_pos[0]=flash_pos[0]+pos[0]*amp;
+        flash_pos[1]=flash_pos[1]+pos[1]*amp;
+        flash_pos[2]=flash_pos[2]+pos[2]*amp;
```
[the remark this answers is not in the document]
Comment (francescopoppi) [reply icaruscode/CRT/CRTPMTMatchingAna_module.cc:492]: This toghether with the previous comment is wonderful.

**Comment by mmrosenberg** [comment]
trigger build LArSoft/lar*@LARSOFT_SUITE_v09_63_00 SBNSoftware/sbncode@v09_63_00_02 SBNSoftware/sbnanaobj@v09_20_00_01 SBNSoftware/sbndaq_artdaq_core@v1_04_02of0 SBNSoftware/sbnobj@v09_15_00_01 SBNSoftware/icarus_signal_processing@v09_63_00_02 SBNSoftware/icarusalg@v09_63_00_02p03 SBNSoftware/icarusutil@v09_62_00

**Comment by PetrilloAtWork** [comment]
I have added a commit in agreement with @francescopoppi.

**Review by PetrilloAtWork — CHANGES_REQUESTED** [review]
A `cout` is left in there that should be removed or replaced.
I have left two more comments (and a commit :roll_eyes:).
The rest is good though.

File: icaruscode/CRT/CRTPMTMatchingAna_module.cc [hunk 3]
```diff
@@ -398,164 +438,174 @@ void icarus::crt::CRTPMTMatchingAna::analyze(art::Event const& e)
...
+      geo::Point_t flash_pos = flashCentroid.middlePoint();
+      t_m = t_m / nPMTsTriggering;
+      if (nPMTsTriggering < 5) {
+        continue;
       }
-      if (m_gate_type==2 || m_gate_type==4){//NuMI OffBeamNuMI
-	if(fRelGateTime>-550 && fRelGateTime<10000) fInTime_gate=true;
-	if(fRelGateTime>-300 && fRelGateTime<9100) fInTime_beam=true;
+      double gateDiff =
+          (Long64_t)m_trigger_timestamp - (Long64_t)m_gate_start_timestamp;
```
[the remark this answers is not in the document]
Comment (francescopoppi) [reply icaruscode/CRT/CRTPMTMatchingAna_module.cc:565]: It is not. Removed redundat variable.

File: icaruscode/CRT/CRTPMTMatchingAna_module.cc [hunk 4]
```diff
@@ -267,127 +299,135 @@ class icarus::crt::CRTPMTMatchingAna : public art::EDAnalyzer {
...
 
-  vector<vector<double>> fTofXYZTHit;   //position/time [cm/ns] for matched OpHit
-  vector<double> fTofXYZTFlash; //position/time [cm/ns] for matched OpFlash
-  vector<vector<double>> fTofXYZTFlashHit;
-  vector<int>            fTofTpcHit;    //TPC for matched OpHit
-  int            fTofTpcFlash;  //TPC for matched OpFlash
+  vector<vector<double>> fTofXYZTHit;   //position/time [cm/ns] for matched
+  OpHit vector<double> fTofXYZTFlash; //position/time [cm/ns] for matched
+  OpFlash vector<vector<double>> fTofXYZTFlashHit; vector<int> fTofTpcHit; //TPC
+  for matched OpHit int            fTofTpcFlash;  //TPC for matched OpFlash
```
[the remark this answers is not in the document]
Comment (francescopoppi) [reply icaruscode/CRT/CRTPMTMatchingAna_module.cc:308]: yes, but on top of that I removed the obsolete and commented tree structure

**Comment by mmrosenberg** [comment]
trigger build LArSoft/lar*@LARSOFT_SUITE_v09_63_00 SBNSoftware/sbncode@v09_63_00_02 SBNSoftware/sbnanaobj@v09_20_00_01 SBNSoftware/sbndaq_artdaq_core@v1_04_02of0 SBNSoftware/sbnobj@v09_15_00_01 SBNSoftware/icarus_signal_processing@v09_63_00_02 SBNSoftware/icarusalg@v09_63_00_02p03 SBNSoftware/icarusutil@v09_62_00

**Review by PetrilloAtWork — APPROVED** [review]
No more comments.
Hey, that's news.

**Pull request merged**